modules/inject/src/main/scala/io/janstenpickle/trace4cats/inject/ContextConstructor.scala

```diff
@@ -0,0 +1,51 @@
+package io.janstenpickle.trace4cats.inject
+
+import cats.Applicative
+import cats.effect.Resource
+
+trait ContextConstructor[F[_], R, R0] { self =>
+  def F: Applicative[F]
+
+  def apply(r: R): Resource[F, R0] = resource(r)
+
+  def resource(r: R): Resource[F, R0]
+
+  def mapBoth[R1](f: (R, R0) => R1): ContextConstructor[F, R, R1] = new ContextConstructor[F, R, R1] {
+    override def F: Applicative[F] = self.F
+
+    override def resource(r: R): Resource[F, R1] = self.resource(r).map(f(r, _))(F)
+  }
+
+  def mapResult[R1](f: R0 => R1): ContextConstructor[F, R, R1] = mapBoth((_, r0) => f(r0))
+
+  def map[R1](f: R => R1): ContextConstructor[F, R, R1] = mapBoth((r, _) => f(r))
+
+  def contramap[R1](f: R1 => R): ContextConstructor[F, R1, R0] = new ContextConstructor[F, R1, R0] {
+    override def F: Applicative[F] = self.F
+
+    override def resource(r: R1): Resource[F, R0] = self.resource(f(r))
+  }
+
+  def mapBothF[R1](f: (R, R0) => F[R1]): ContextConstructor[F, R, R1] = new ContextConstructor[F, R, R1] {
+    override def F: Applicative[F] = self.F
+
+    override def resource(r: R): Resource[F, R1] = self.resource(r).evalMap(f(r, _))(F)
+  }
+
+  def mapBothResource[R1](f: (R, R0) => Resource[F, R1]): ContextConstructor[F, R, R1] =
+    new ContextConstructor[F, R, R1] {
+      override def F: Applicative[F] = self.F
+
+      override def resource(r: R): Resource[F, R1] = self.resource(r).flatMap(f(r, _))
+    }
+}
+
+object ContextConstructor {
+
+  def instance[F[_]: Applicative, R, R0](f: R => Resource[F, R0]): ContextConstructor[F, R, R0] =
+    new ContextConstructor[F, R, R0] {
+      override def F: Applicative[F] = implicitly
+
+      override def resource(r: R): Resource[F, R0] = f(r)
+    }
+}
```

Comment:
Here's my draft implementation of the same trait. I renamed it to let you import it without issues and try it instead of yours.

```scala
package io.janstenpickle.trace4cats.inject

import cats.Applicative
import cats.effect.Resource

trait KontextConstructor[F[_], -In, +Out] { self =>
  implicit protected def F: Applicative[F]

  @inline def apply(in: In): Resource[F, Out] = resource(in)

  def resource(in: In): Resource[F, Out]

  // in functor notation
  def map[Out1](f: Out => Out1): KontextConstructor[F, In, Out1] = KontextConstructor(resource(_).map(f))
  def contramap[In1](f: In1 => In): KontextConstructor[F, In1, Out] = KontextConstructor(in1 => resource(f(in1)))

  
  // `evalMap` in resource and stream notations, aka `semiflatMap` in transformers, aka `flatMapF` in Kleisli
  def evalMap[Out1](f: Out => F[Out1]): KontextConstructor[F, In, Out1] = KontextConstructor(
    self.resource(_).evalMap(f)
  )

  def andThen[Out1](f: Out => Resource[F, Out1]): KontextConstructor[F, In, Out1] = KontextConstructor(
    resource(_).flatMap(f)
  )
  def andThen[Out1](f: KontextConstructor[F, Out, Out1]): KontextConstructor[F, In, Out1] = andThen(f.resource _)

  def compose[In1](f: In1 => Resource[F, In]): KontextConstructor[F, In1, Out] = KontextConstructor(f).andThen(self)
  def compose[In1](f: KontextConstructor[F, In1, In]): KontextConstructor[F, In1, Out] = f.andThen(self)

  // aka `merge id` in arrow notation
  def withInput[In1 <: In]: KontextConstructor[F, In1, (In1, Out)] =
    KontextConstructor(in1 => resource(in1).map(out => (in1, out)))
}

object KontextConstructor {
  def apply[F[_], In, Out](f: In => Resource[F, Out])(implicit F0: Applicative[F]): KontextConstructor[F, In, Out] =
    new KontextConstructor[F, In, Out] {
      def F: Applicative[F] = F0
      def resource(in: In): Resource[F, Out] = f(in)
    }
}
```

Reply:
Nice, thanks for this 👍